gcp/modules/gcp-stackdriver-monitoring/alert_disk_snapshots.tf

```diff
@@ -0,0 +1,79 @@
+locals {
+  pv_name_regexp = "^pv-database-storage-couchdb-couchdb-[0-9]+$"
+}
+
+resource "google_monitoring_alert_policy" "disk_snapshots" {
+  display_name = "Snapshots are being created for all persistent volumes"
+  combiner     = "OR"
+
+  conditions = [
+    {
+      condition_absent {
+        filter   = "metric.type=\"logging.googleapis.com/user/compute.disks.createSnapshot\" resource.type=\"gce_disk\" AND metric.label.pv_name=monitoring.regex.full_match(\"${local.pv_name_regexp}\") AND metric.label.severity=\"NOTICE\""
+        duration = "600s"
+
+        aggregations {
+          alignment_period     = "300s"
+          per_series_aligner   = "ALIGN_SUM"
+          cross_series_reducer = "REDUCE_SUM"
+
+          group_by_fields = [
+            "metric.label.pv_name",
+          ]
+        }
+      }
+
+      display_name = "Snapshot creation events are missing for some persistent volumes"
+    },
+    {
+      condition_threshold {
+        filter = "metric.type=\"logging.googleapis.com/user/compute.disks.createSnapshot\" resource.type=\"gce_disk\" AND metric.label.pv_name=monitoring.regex.full_match(\"${local.pv_name_regexp}\") AND metric.label.severity=\"NOTICE\""
+
+        comparison      = "COMPARISON_LT"
+        threshold_value = 1.0
+        duration        = "600s"
+
+        aggregations {
+          alignment_period     = "300s"
+          per_series_aligner   = "ALIGN_SUM"
+          cross_series_reducer = "REDUCE_SUM"
+
+          group_by_fields = [
+            "metric.label.pv_name",
+          ]
+        }
+      }
+
+      display_name = "Snapshot creation events are occurring too infrequently for some persistent volumes"
+    },
+    {
+      condition_threshold {
+        filter = "metric.type=\"logging.googleapis.com/user/compute.disks.createSnapshot\" resource.type=\"gce_disk\" AND metric.label.severity=\"ERROR\""
+
+        aggregations {
+          alignment_period   = "300s"
+          per_series_aligner = "ALIGN_SUM"
+        }
+
+        denominator_filter = "metric.type=\"logging.googleapis.com/user/compute.disks.createSnapshot\" resource.type=\"gce_disk\" AND metric.label.severity!=\"ERROR\""
+
+        denominator_aggregations {
+          alignment_period   = "300s"
+          per_series_aligner = "ALIGN_SUM"
+        }
+
+        comparison      = "COMPARISON_GT"
+        threshold_value = 0.05
+        duration        = "0s"
```

Comment:
Shouldn't this duration be higher than `0s`, to prevent flapping?

I would consider setting it to `600s` to be consistent with the other alerts, that way it would cover more than 1 value and only alert if the issue is not resolved. Otherwise, this can be triggered by a single `ERROR` event.

Reply:
Policy will be triggered in case amount of errors exceed 5% of total amount of snapshots for the alignment period. But in case total amount of snapshots is low, of course the single error event is enough. I don't think increasing duration is desirable here, because in this case policy will be ignoring short error bursts, so I would increase threshold and alignment window instead if we want it to be less sensitive.

Reply:
This is the same problem as already discussed if `duration` does not cover more than one window.

Let's take an example of a single error, and assume snapshots are suppoed to happen every 5m:
```
|1|2|3|4|5|6|7|8|9|0|1|2|3|4|5|6|7|8|9|0| < time
|-|-|-|-|-|-|-|-|-|-|-|-|-|-|-|-|-|-|-|-|
|1|_|_|_|_|_|_|_|_|_|_|_|_|_|_|_|_|_|_|_| < incoming data timeseries (errors)
|1|_|_|_|_|1|_|_|_|_|1|_|_|_|_|1|_|_|_|_| < incoming data timeseries (snapshots)
 [-----------------] [-----------------]  < alignment blocks 10m
|_|_|_|_|_|_|_|_|_|5|_|_|_|_|_|_|_|_|_|0| < aligned data after denominator (in 10s of %)
                    ^ alert triggers immediately
                                        ^ alert is recovered
```
^ with `0` duration and `10m` alignment, single error triggered the alert, 5m after there were successful snapshots without any further errors, the alert is delayed 5m and recovery basically 15m.


```
|1|2|3|4|5|6|7|8|9|0|1|2|3|4|5|6|7|8|9|0| < time
|-|-|-|-|-|-|-|-|-|-|-|-|-|-|-|-|-|-|-|-|
|1|_|_|_|_|_|_|_|_|_|_|_|_|_|_|_|_|_|_|_| < incoming data timeseries (errors)
|1|_|_|_|_|1|_|_|_|_|1|_|_|_|_|1|_|_|_|_| < incoming data timeseries (snaprshots)
 [-------] [-------] [-------] [-------]  < alignment blocks 10m
|_|_|_|_|10_|_|_|_|0|_|_|_|_|0|_|_|_|_|0| < aligned data after denominator (in 10s of %)
```
^ with `10m` duration and `5m` alignment, a single error did not trigger the alert (if there would be 2 consecutive errors, the alert would be triggered)

Now there might be another issue (and I believe there is) with how SD treats metrics in absence of data. I believe that if either of `filter` or `denominator_filter` results in no metrics, SD will not produce a data point. Now when SD evaluates alerts, it always looks at last present datapoint. This would mean that in case of error, these alerts are never recovered (imagine `empty` space instead of `0` in the above examples). So each alert would get "stuck" as we would only get data points when there's an error. I'm not sure how resolve this atm., but I'll think about it a bit more.

Reply:
There are obvious disadvantages to both approaches: with the `duration` set to include at least two data points, condition would ignore short error bursts:
```
|1|2|3|4|5|6|7|8|9|0|1|2|3|4|5|6|7|8|9|0| < time
|-|-|-|-|-|-|-|-|-|-|-|-|-|-|-|-|-|-|-|-|
|_|_|_|_|1|_|_|_|_|_|_|_|_|_|1|_|_|_|_|_| < incoming data timeseries (errors)
|1|_|_|_|_|1|_|_|_|_|_|_|_|_|_|1|_|_|_|_| < incoming data timeseries (snapshots)
 [-------] [-------] [-------] [-------]  < alignment blocks 5m
|_|_|_|_|100|_|_|_|_|0|_|_|_|_|100|_|_|_|_|0| < aligned data after denominator
         ^ metric is above threshold, 100% snapshots failed
         		   ^ metric recovered within duration interval, no alert triggered
```

And yes, with `duration` set to 0, condition is more sensitive, when overall number of events is low per alignment period. 

I am still slightly in favor of latter approach, but I also feel fair amount of over-engineering here – remember, policy deployed to my dev env and I observe 0 false positives for 1+ week. So I think it is fine to deploy this as-is, we can fine tune some parameters later if needed. 

> Now there might be another issue (and I believe there is) with how SD treats metrics in absence of data

Yeah, I remember we bumped into this "feature" of Stackdriver couple of times with other policies. I'll try to reproduce this to confirm, but not sure what can be done.

Reply:
>  condition would ignore short error bursts

I think this would be desirable, if the errors stop happening I don't think we want to care